docs: Move Localnet development to separate page (#977)
oasisprotocol/docs: docs/dapp/tools/localnet.mdx

Private Keys
==================
(0) 0x617346c545d62b8213ea907acf1b570a7405683e2c6dcaf963fc21fd677e0c56
(1) 0xf82d6e09208b0bd44a397f7e73b05c564e6c9f70b151ee7677e2bb8d6ce5d882
(2) 0xeb2f21d20086f3dd6bfe7184dad1cb8b0fb802f27b1334e836a19eda0a43a1c2
(3) 0x82b0203d6063992b1052004b90411c45d4f3afab696346f006e74c6abd8f855e
(4) 0x7179c6e1add3a2993822653b9c98fe606f47fb6d4c0d0d81b31b067cf6bb5f83

HD Wallet
==================
Mnemonic:       coach genre beach child crunch champion tell adult critic peace canoe stable
Base HD Path:   m/44'/60'/0'/0/%d

WARNING: The chain is running in ephemeral mode. State will be lost after restart!

 * Listening on http://localhost:8545 and ws://localhost:8546. Chain ID: 0x5afd
 * Container start-up took 66 seconds, node log level is set to warn.
```

Those familiar with local dApp environments will find the output above similar
to `geth --dev` or `ganache-cli` commands or the `geth-dev-assistant` npm
package. The [sapphire-localnet] will spin up a private Oasis Network locally,
generate and populate test accounts and make the following Web3 endpoints
available for you to use:

- `http://localhost:8545`
- `ws://localhost:8546`

> 

Reviewer: ```suggestion
Additionally, the [Oasis GRPC](https://github.com/oasisprotocol/oasis-core/blob/master/docs/oasis-node/rpc.md) endpoint is exposed on:

- `http://localhost:8544`

```

> :::tip
> 
> If you prefer using the same mnemonics each time (e.g. for testing purposes)
> or to populate just a single account, use `-to` flag and pass the mnemonics or
> the wallet addresses. By passing the `-test-mnemonic` flag you can fund the
> standard test accounts provided by the `hardhat node` commmand and that are
> typically used for solidity unit tests.
> 
> <Tabs>
> <TabItem value="Sapphire">
> 
> ```sh
> docker run -it -p8544-8546:8544-8546 ghcr.io/oasisprotocol/sapphire-localnet -to "bench remain brave curve frozen verify dream margin alarm world repair innocent" -n3
> docker run -it -p8544-8546:8544-8546 ghcr.io/oasisprotocol/sapphire-localnet -to "0x75eCF0d4496C2f10e4e9aF3D4d174576Ee9010E2,0xbDA5747bFD65F08deb54cb465eB87D40e51B197E"
> docker run -it -p8544-8546:8544-8546 ghcr.io/oasisprotocol/sapphire-localnet -test-mnemonic
> ```
> 
> </TabItem>
> <TabItem value="Emerald">
> 
> ```sh
> docker run -it -p8544-8546:8544-8546 ghcr.io/oasisprotocol/emerald-localnet -to "bench remain brave curve frozen verify dream margin alarm world repair innocent" -n3
> docker run -it -p8544-8546:8544-8546 ghcr.io/oasisprotocol/emerald-localnet -to "0x75eCF0d4496C2f10e4e9aF3D4d174576Ee9010E2,0xbDA5747bFD65F08deb54cb465eB87D40e51B197E"
> docker run -it -p8544-8546:8544-8546 ghcr.io/oasisprotocol/emerald-localnet -test-mnemonic
> ```
> 
> </TabItem>
> </Tabs>
> 
> :::

Reviewer: The `-test-mnemonic` is now default. Let's keep this outside of admonition not to bloat the chapter too much.

````suggestion
By default, the Localnet docker image will populate the first five accounts
derived from the standard test mnemonic, compatible with `hardhat node`.
These accounts are typically used for solidity unit tests. If you prefer
populating different accounts, use `-to` flag and pass the mnemonics or the
wallet addresses. Using the `-n` parameter defines the number of derived
addresses to fund.

<Tabs>
<TabItem value="Sapphire">

```sh
docker run -it -p8544-8546:8544-8546 ghcr.io/oasisprotocol/sapphire-localnet -to "bench remain brave curve frozen verify dream margin alarm world repair innocent" -n3
docker run -it -p8544-8546:8544-8546 ghcr.io/oasisprotocol/sapphire-localnet -to "0x75eCF0d4496C2f10e4e9aF3D4d174576Ee9010E2,0xbDA5747bFD65F08deb54cb465eB87D40e51B197E"
```

</TabItem>
<TabItem value="Emerald">

```sh
docker run -it -p8544-8546:8544-8546 ghcr.io/oasisprotocol/emerald-localnet -to "bench remain brave curve frozen verify dream margin alarm world repair innocent" -n3
docker run -it -p8544-8546:8544-8546 ghcr.io/oasisprotocol/emerald-localnet -to "0x75eCF0d4496C2f10e4e9aF3D4d174576Ee9010E2,0xbDA5747bFD65F08deb54cb465eB87D40e51B197E"
```

</TabItem>
</Tabs>
````

:::note Running on Apple M chips

There is currently no `arm64` build available for M Macs, so you will need to
force the docker image to use the `linux/x86_64` platform, like this:

<Tabs>
<TabItem value="Sapphire">

```sh
docker run -it -p8544-8546:8544-8546 --platform linux/x86_64 ghcr.io/oasisprotocol/sapphire-localnet
```

</TabItem>
<TabItem value="Emerald">

```sh
docker run -it -p8544-8546:8544-8546 --platform linux/x86_64 ghcr.io/oasisprotocol/emerald-localnet
```

</TabItem>
</Tabs>

:::

:::danger

[sapphire-localnet] and [emerald-localnet] run in ephemeral mode. Any smart
contract and wallet balance will be lost after you quit the Docker container!

:::

[sapphire-localnet]: https://github.com/oasisprotocol/oasis-web3-gateway/pkgs/container/sapphire-localnet

> [emerald-localnet]: https://github.com/oasisprotocol/oasis-web3-gateway/pkgs/container/emerald-localnet

Reviewer: Please also add a section on how to run Localnet with github actions. For example:

```yaml
  jobs:
    services:
      sapphire-localnet-ci:
        image: ghcr.io/oasisprotocol/sapphire-localnet
        ports:
          - 8544:8544
          - 8545:8545
          - 8546:8546
        env:
          OASIS_DEPOSIT_BINARY: /oasis-deposit -n 2
        options: >-
          --rm
          --health-cmd="test -f /CONTAINER_READY"
          --health-start-period=90s
```

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Localnet

For convenient development and testing of your dApps the Oasis team prepared
both [ghcr.io/oasisprotocol/sapphire-localnet][sapphire-localnet] and
[ghcr.io/oasisprotocol/emerald-localnet][emerald-localnet] Docker images.

> They will bring you a complete Oasis network stack to your desktop. The

Author: ```suggestion
They will bring you a complete Oasis network stack to your workspace. The
```

Localnet Sapphire instance **mimics confidential transactions**, but it does
not run in a trusted execution environment nor does it require Intel's SGX on
your computer. The network is isolated from the Mainnet or Testnet and consists
of a:

- single Oasis validator node with 1-second block time and 30-second epoch,
- single Oasis client node,
- single compute node running Oasis Sapphire or Emerald,
- single key manager node,
- PostgreSQL instance,
- Oasis Web3 gateway with transaction indexer and enabled Oasis RPCs,
- helper script which populates the account(s) for you.

:::note Hardware requirements

You will need at least 16GB of RAM to run the Docker images in addition to your
machine's OS.

:::

To run the image, execute:

<Tabs>
<TabItem value="Sapphire">

```sh
docker run -it -p8544-8546:8544-8546 ghcr.io/oasisprotocol/sapphire-localnet
```

</TabItem>
<TabItem value="Emerald">

```sh
docker run -it -p8544-8546:8544-8546 ghcr.io/oasisprotocol/emerald-localnet
```

</TabItem>
</Tabs>

After a while, running the `sapphire-localnet` will show you something like:

```

> sapphire-localnet 2024-05-28-git37b7166 (oasis-core: 24.0-gitfb49717, sapphire-paratime: 0.7.3-testnet, oasis-web3-gateway: 5.1.0)

Author: I don't think devs need duplicated info for Emerald.

Reviewer 2: I think a output with a newer docker version would be good.
In this regard some info to update the docker image, with `docker pull ghcr.io/oasisprotocol/sapphire-localnet:latest` to get the latest version which includes ROFL, could be helpful for devs that used localnet already months ago.

HD Wallet
==================
Mnemonic:       coach genre beach child crunch champion tell adult critic peace canoe stable
Base HD Path:   m/44'/60'/0'/0/%d

WARNING: The chain is running in ephemeral mode. State will be lost after restart!

 * Listening on http://localhost:8545 and ws://localhost:8546. Chain ID: 0x5afd
 * Container start-up took 66 seconds, node log level is set to warn.
```

Those familiar with local dApp environments will find the output above similar
to `geth --dev` or `ganache-cli` commands or the `geth-dev-assistant` npm
package. The [sapphire-localnet] will spin up a private Oasis Network locally,
generate and populate test accounts and make the following Web3 endpoints
available for you to use:

- `http://localhost:8545`
- `ws://localhost:8546`

:::tip

If you prefer using the same mnemonics each time (e.g. for testing purposes)
or to populate just a single account, use `-to` flag and pass the mnemonics or
the wallet addresses. By passing the `-test-mnemonic` flag you can fund the
standard test accounts provided by the `hardhat node` commmand and that are
typically used for solidity unit tests.

> 

Author: ```suggestion
```

(4) 0x7179c6e1add3a2993822653b9c98fe606f47fb6d4c0d0d81b31b067cf6bb5f83

HD Wallet
==================
Mnemonic:       coach genre beach child crunch champion tell adult critic peace canoe stable
Base HD Path:   m/44'/60'/0'/0/%d

WARNING: The chain is running in ephemeral mode. State will be lost after restart!

 * Listening on http://localhost:8545 and ws://localhost:8546. Chain ID: 0x5afd
 * Container start-up took 66 seconds, node log level is set to warn.
```

Those familiar with local dApp environments will find the output above similar
to `geth --dev` or `ganache-cli` commands or the `geth-dev-assistant` npm
package. The [sapphire-localnet] will spin up a private Oasis Network locally,
generate and populate test accounts and make the following Web3 endpoints
available for you to use:

- `http://localhost:8545`
- `ws://localhost:8546`

Additionally, the [Oasis GRPC][oasis-rpc] endpoint is exposed on:

- `http://localhost:8544`

[oasis-rpc]: https://github.com/oasisprotocol/oasis-core/blob/master/docs/oasis-node/rpc.md

By default, the Localnet docker image will populate the first five accounts
derived from the standard test mnemonic, compatible with `hardhat node`.

> These accounts are typically used for solidity unit tests. If you prefer

Author: ```suggestion
These accounts are typically used for Solidity unit tests. If you prefer
```

populating different accounts, use `-to` flag and pass the mnemonics or the

> wallet addresses. Using the `-n` parameter defines the number of derived

Author: ```suggestion
wallet addresses. Use the `-n` parameter to define the number of derived
```

docker run -it -p8544-8546:8544-8546 ghcr.io/oasisprotocol/emerald-localnet -to "0x75eCF0d4496C2f10e4e9aF3D4d174576Ee9010E2,0xbDA5747bFD65F08deb54cb465eB87D40e51B197E"
```

</TabItem>
</Tabs>

:::note Running on Apple M chips

There is currently no `arm64` build available for M Macs, so you will need to
force the docker image to use the `linux/x86_64` platform, like this:

<Tabs>
<TabItem value="Sapphire">

```sh
docker run -it -p8544-8546:8544-8546 --platform linux/x86_64 ghcr.io/oasisprotocol/sapphire-localnet
```

</TabItem>
<TabItem value="Emerald">

```sh
docker run -it -p8544-8546:8544-8546 --platform linux/x86_64 ghcr.io/oasisprotocol/emerald-localnet
```

</TabItem>
</Tabs>

:::

:::danger

[sapphire-localnet] and [emerald-localnet] run in ephemeral mode. Any smart
contract and wallet balance will be lost after you quit the Docker container!

:::

[sapphire-localnet]: https://github.com/oasisprotocol/oasis-web3-gateway/pkgs/container/sapphire-localnet
[emerald-localnet]: https://github.com/oasisprotocol/oasis-web3-gateway/pkgs/container/emerald-localnet

## GitHub Actions

You can easily integrate a Sapphire localnet into your CI/CD workflow. Use the

> example GitHub Action configuration to start a Sapphire stack and expose the
> necessary ports for testing.

Reviewer: ```suggestion
example GitHub Action configuration to start a Sapphire stack, fund the first two test
accounts and expose the necessary ports for testing.
```

> ```yaml

Reviewer: Also add a tab for Emerald config for consistency.
